File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.firestore.testutil;

import static com.google.firebase.firestore.testutil.TestUtil.map;
import static com.google.firebase.firestore.util.Util.autoId;
import static java.util.Arrays.asList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import android.content.Context;
import android.os.StrictMode;
import androidx.test.core.app.ApplicationProvider;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.firestore.AccessHelper;
import com.google.firebase.firestore.BuildConfig;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreSettings;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.MetadataChanges;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.firestore.Source;
import com.google.firebase.firestore.WriteBatch;
import com.google.firebase.firestore.auth.User;
import com.google.firebase.firestore.core.DatabaseInfo;
import com.google.firebase.firestore.model.DatabaseId;
import com.google.firebase.firestore.testutil.provider.FirestoreProvider;
import com.google.firebase.firestore.util.AsyncQueue;
import com.google.firebase.firestore.util.Listener;
import com.google.firebase.firestore.util.Logger;
import com.google.firebase.firestore.util.Logger.Level;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

class MockCredentialsProvider extends EmptyCredentialsProvider {

  private static MockCredentialsProvider instance;
  private Listener<User> listener;

  public static MockCredentialsProvider instance() {
    if (MockCredentialsProvider.instance == null) {
      MockCredentialsProvider.instance = new MockCredentialsProvider();
    }
    return MockCredentialsProvider.instance;
  }

  private MockCredentialsProvider() {}

  @Override
  public void setChangeListener(Listener<User> changeListener) {
    super.setChangeListener(changeListener);
    this.listener = changeListener;
  }

  public void changeUserTo(User user) {
    listener.onValue(user);
  }
}

/** A set of helper methods for tests */
public class IntegrationTestUtil {

  public enum TargetBackend {
    EMULATOR,
    QA,
    NIGHTLY,
    PROD
  }

  // Set this to the desired enum value to change the target backend when running tests locally.
  // Note: DO NOT change this variable except for local testing.
  private static final TargetBackend backendForLocalTesting = null;

  private static final TargetBackend backend = getTargetBackend();
  private static final String EMULATOR_HOST = "10.0.2.2";
  private static final int EMULATOR_PORT = 8080;

  // Alternate project ID for creating "bad" references. Doesn't actually need to work.
  public static final String BAD_PROJECT_ID = "test-project-2";

  /** Online status of all active Firestore clients. */
  private static final Map<FirebaseFirestore, Boolean> firestoreStatus = new HashMap<>();

  /** Default amount of time to wait for a given operation to complete, used by waitFor() helper. */
  private static final long OPERATION_WAIT_TIMEOUT_MS = 30000;

  /**
   * Firestore databases can be subject to a ~30s "cold start" delay if they have not been used
   * recently, so before any tests run we "prime" the backend.
   */
  private static final long PRIMING_TIMEOUT_MS = 45000;

  private static final FirestoreProvider provider = new FirestoreProvider();

  private static boolean strictModeEnabled = false;
  private static boolean backendPrimed = false;

  // FirebaseOptions needed to create a test FirebaseApp.
  private static final FirebaseOptions OPTIONS =
      new FirebaseOptions.Builder()
          .setApplicationId(":123:android:123ab")
          .setProjectId(provider.projectId())
          .build();

  public static FirestoreProvider provider() {
    return provider;
  }

  private static String getFirestoreHost() {
    switch (backend) {
      case EMULATOR:
        return String.format("%s:%d", EMULATOR_HOST, EMULATOR_PORT);
      case QA:
        return "staging-firestore.sandbox.googleapis.com";
      case NIGHTLY:
        return "test-firestore.sandbox.googleapis.com";
      case PROD:
      default:
        return "firestore.googleapis.com";
    }
  }

  private static boolean getSslEnabled() {
    // ssl is enabled in all environments except for the emulator.
    return !isRunningAgainstEmulator();
  }

  public static TargetBackend getTargetBackend() {
    if (backendForLocalTesting != null) {
      return backendForLocalTesting;
    }
    switch (BuildConfig.TARGET_BACKEND) {
      case "emulator":
        return TargetBackend.EMULATOR;
      case "qa":
        return TargetBackend.QA;
      case "nightly":
        return TargetBackend.NIGHTLY;
      case "prod":
        return TargetBackend.PROD;
      default:
        throw new RuntimeException("Unknown backend configuration used for integration tests.");
    }
  }

  public static DatabaseInfo testEnvDatabaseInfo() {
    return new DatabaseInfo(
        DatabaseId.forProject(provider.projectId()),
        "test-persistenceKey",
        getFirestoreHost(),
        getSslEnabled());
  }

  public static FirebaseFirestoreSettings newTestSettings() {
    Logger.debug("IntegrationTestUtil", "target backend is: %s", backend.name());
    FirebaseFirestoreSettings.Builder settings = new FirebaseFirestoreSettings.Builder();
    settings.setHost(getFirestoreHost());
    settings.setSslEnabled(getSslEnabled());
    return settings.build();
  }

  public static FirebaseApp testFirebaseApp() {
    try {
      return FirebaseApp.getInstance(FirebaseApp.DEFAULT_APP_NAME);
    } catch (IllegalStateException e) {
      return FirebaseApp.initializeApp(ApplicationProvider.getApplicationContext(), OPTIONS);
    }
  }

  /** Initializes a new Firestore instance that uses the default project. */
  public static FirebaseFirestore testFirestore() {
    return testFirestore(newTestSettings());
  }

  /**
   * Initializes a new Firestore instance that uses the default project, customized with the
   * provided settings.
   */
  public static FirebaseFirestore testFirestore(FirebaseFirestoreSettings settings) {
    FirebaseFirestore firestore = testFirestore(provider.projectId(), Level.DEBUG, settings);
    primeBackend();
    return firestore;
  }

  private static void primeBackend() {
    if (!backendPrimed) {
      backendPrimed = true;
      TaskCompletionSource<Void> watchInitialized = new TaskCompletionSource<>();
      TaskCompletionSource<Void> watchUpdateReceived = new TaskCompletionSource<>();
      DocumentReference docRef = testDocument();
      ListenerRegistration listenerRegistration =
          docRef.addSnapshotListener(
              (snapshot, error) -> {
                assertNull(error);
                if ("done".equals(snapshot.get("value"))) {
                  watchUpdateReceived.setResult(null);
                } else {
                  watchInitialized.setResult(null);
                }
              });

      // Wait for watch to initialize and deliver first event.
      waitFor(watchInitialized.getTask());

      // Use a transaction to perform a write without triggering any local events.
      docRef
          .getFirestore()
          .runTransaction(
              transaction -> {
                transaction.set(docRef, map("value", "done"));
                return null;
              });

      // Wait to see the write on the watch stream.
      waitFor(watchUpdateReceived.getTask(), PRIMING_TIMEOUT_MS);

      listenerRegistration.remove();
    }
  }

  /** Initializes a new Firestore instance that uses a non-existing default project. */
  public static FirebaseFirestore testAlternateFirestore() {
    return testFirestore(BAD_PROJECT_ID, Level.DEBUG, newTestSettings());
  }

  /**
   * Enable strict mode for integration tests. Currently checks for leaked SQLite or other Closeable
   * objects.
   *
   * <p>If a leak is found, Android will log the leak and kill the test.
   */
  private static void ensureStrictMode() {
    if (strictModeEnabled) {
      return;
    }

    strictModeEnabled = true;
    StrictMode.setVmPolicy(
        new StrictMode.VmPolicy.Builder()
            .detectLeakedSqlLiteObjects()
            .detectLeakedClosableObjects()
            .penaltyLog()
            .penaltyDeath()
            .build());
  }

  /**
   * Initializes a new Firestore instance that can be used in testing. It is guaranteed to not share
   * state with other instances returned from this call.
   */
  public static FirebaseFirestore testFirestore(
      String projectId, Logger.Level logLevel, FirebaseFirestoreSettings settings) {
    String persistenceKey = "db" + firestoreStatus.size();
    return testFirestore(projectId, logLevel, settings, persistenceKey);
  }

  public static FirebaseFirestore testFirestore(
      String projectId,
      Logger.Level logLevel,
      FirebaseFirestoreSettings settings,
      String persistenceKey) {
    return testFirestore(
        DatabaseId.forDatabase(projectId, BuildConfig.TARGET_DATABASE_ID),
        logLevel,
        settings,
        persistenceKey);
  }

  public static FirebaseFirestore testFirestore(
      DatabaseId databaseId,
      Logger.Level logLevel,
      FirebaseFirestoreSettings settings,
      String persistenceKey) {
    // This unfortunately is a global setting that affects existing Firestore clients.
    Logger.setLogLevel(logLevel);

    Context context = ApplicationProvider.getApplicationContext();

    ensureStrictMode();

    AsyncQueue asyncQueue = new AsyncQueue();

    FirebaseFirestore firestore =
        AccessHelper.newFirebaseFirestore(
            context,
            databaseId,
            persistenceKey,
            MockCredentialsProvider.instance(),
            new EmptyAppCheckTokenProvider(),
            asyncQueue,
            /*firebaseApp=*/ null,
            /*instanceRegistry=*/ (dbId) -> {});
    waitFor(firestore.clearPersistence());
    firestore.setFirestoreSettings(settings);
    firestoreStatus.put(firestore, true);

    return firestore;
  }

  public static void tearDown() {
    try {
      for (FirebaseFirestore firestore : firestoreStatus.keySet()) {
        Task<Void> result = firestore.terminate();
        waitFor(result);
      }
    } finally {
      firestoreStatus.clear();
    }
  }

  public static DocumentReference testDocument() {
    return testCollection("test-collection").document();
  }

  public static DocumentReference testDocumentWithData(Map<String, Object> data) {
    DocumentReference docRef = testDocument();
    waitFor(docRef.set(data));
    return docRef;
  }

  public static CollectionReference testCollection() {
    return testFirestore().collection(autoId());
  }

  public static CollectionReference testCollection(String name) {
    return testFirestore().collection(name + "_" + autoId());
  }

  public static CollectionReference testCollectionWithDocs(Map<String, Map<String, Object>> docs) {
    CollectionReference collection = testCollection();
    CollectionReference writer = testFirestore().collection(collection.getId());
    writeAllDocs(writer, docs);
    return collection;
  }

  public static void writeAllDocs(
      CollectionReference collection, Map<String, Map<String, Object>> docs) {
    WriteBatch writeBatch = null;
    int writeBatchSize = 0;

    for (Map.Entry<String, Map<String, Object>> doc : docs.entrySet()) {
      if (writeBatch == null) {
        writeBatch = collection.getFirestore().batch();
      }

      writeBatch.set(collection.document(doc.getKey()), doc.getValue());
      writeBatchSize++;

      // Write batches are capped at 500 writes. Use 400 just to be safe.
      if (writeBatchSize == 400) {
        waitFor(writeBatch.commit());
        writeBatch = null;
        writeBatchSize = 0;
      }
    }

    if (writeBatch != null) {
      waitFor(writeBatch.commit());
    }
  }

  public static void waitForOnlineSnapshot(DocumentReference doc) {
    TaskCompletionSource<Void> done = new TaskCompletionSource<>();
    ListenerRegistration registration =
        doc.addSnapshotListener(
            MetadataChanges.INCLUDE,
            (snapshot, error) -> {
              assertNull(error);
              if (!snapshot.getMetadata().isFromCache()) {
                done.setResult(null);
              }
            });
    waitFor(done.getTask());
    registration.remove();
  }

  public static void waitFor(Semaphore semaphore) {
    waitFor(semaphore, 1);
  }

  public static void waitFor(Semaphore semaphore, int count) {
    try {
      boolean acquired =
          semaphore.tryAcquire(count, OPERATION_WAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
      if (!acquired) {
        throw new TimeoutException("Failed to acquire semaphore within test timeout");
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  public static void waitFor(CountDownLatch countDownLatch) {
    try {
      boolean acquired = countDownLatch.await(OPERATION_WAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
      if (!acquired) {
        throw new TimeoutException("Failed to acquire countdown latch within test timeout");
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> T waitFor(Task<T> task) {
    return waitFor(task, OPERATION_WAIT_TIMEOUT_MS);
  }

  public static <T> T waitFor(Task<T> task, long timeoutMS) {
    try {
      return Tasks.await(task, timeoutMS, TimeUnit.MILLISECONDS);
    } catch (TimeoutException | ExecutionException | InterruptedException e) {
      throw new RuntimeException(e);
    }
  }

  public static <T> Exception waitForException(Task<T> task) {
    try {
      Tasks.await(task, OPERATION_WAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
      throw new RuntimeException("Expected Exception but Task completed successfully.");
    } catch (ExecutionException e) {
      return (Exception) e.getCause();
    } catch (TimeoutException | InterruptedException e) {
      throw new RuntimeException(e);
    }
  }

  public static List<Map<String, Object>> querySnapshotToValues(QuerySnapshot querySnapshot) {
    List<Map<String, Object>> res = new ArrayList<>();
    for (DocumentSnapshot doc : querySnapshot) {
      res.add(doc.getData());
    }
    return res;
  }

  public static List<String> querySnapshotToIds(QuerySnapshot querySnapshot) {
    List<String> res = new ArrayList<>();
    for (DocumentSnapshot doc : querySnapshot) {
      res.add(doc.getId());
    }
    return res;
  }

  public static void disableNetwork(FirebaseFirestore firestore) {
    if (firestoreStatus.get(firestore)) {
      waitFor(firestore.disableNetwork());
      firestoreStatus.put(firestore, false);
    }
  }

  public static void enableNetwork(FirebaseFirestore firestore) {
    if (!firestoreStatus.get(firestore)) {
      waitFor(firestore.enableNetwork());
      // Wait for the client to connect.
      waitFor(firestore.collection("unknown").document().delete());
      firestoreStatus.put(firestore, true);
    }
  }

  public static boolean isNetworkEnabled(FirebaseFirestore firestore) {
    return firestoreStatus.get(firestore);
  }

  public static void removeFirestore(FirebaseFirestore firestore) {
    firestoreStatus.remove(firestore);
  }

  public static Map<String, Object> toDataMap(QuerySnapshot qrySnap) {
    Map<String, Object> result = new HashMap<>();
    for (DocumentSnapshot docSnap : qrySnap.getDocuments()) {
      result.put(docSnap.getId(), docSnap.getData());
    }
    return result;
  }

  public static boolean isRunningAgainstEmulator() {
    return backend.equals(TargetBackend.EMULATOR);
  }

  public static void testChangeUserTo(User user) {
    MockCredentialsProvider.instance().changeUserTo(user);
  }

  public static List<Object> nullList() {
    List<Object> nullArray = new ArrayList<>();
    nullArray.add(null);
    return nullArray;
  }

  /**
   * Checks that running the query while online (against the backend/emulator) results in the same
   * documents as running the query while offline. If `expectedDocs` is provided, it also checks
   * that both online and offline query result is equal to the expected documents.
   *
   * @param query The query to check
   * @param expectedDocs Ordered list of document keys that are expected to match the query
   */
  public static void checkOnlineAndOfflineResultsMatch(Query query, String... expectedDocs) {
    QuerySnapshot docsFromServer = waitFor(query.get(Source.SERVER));
    QuerySnapshot docsFromCache = waitFor(query.get(Source.CACHE));

    assertEquals(querySnapshotToIds(docsFromServer), querySnapshotToIds(docsFromCache));
    List<String> expected = asList(expectedDocs);
    if (!expected.isEmpty()) {
      assertEquals(expected, querySnapshotToIds(docsFromCache));
    }
  }
}

File: firebase-firestore/ktx/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.kt
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.firestore.testutil

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.google.firebase.firestore.Transaction
import com.google.firebase.firestore.UserDataReader
import com.google.firebase.firestore.ktx.BuildConfig
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.firestore.ktx.firestoreSettings
import com.google.firebase.firestore.ktx.serialization.decodeFromMap
import com.google.firebase.firestore.ktx.serialization.encodeToMap
import com.google.firebase.firestore.model.DocumentKey
import com.google.firebase.firestore.model.mutation.FieldMask
import com.google.firebase.firestore.util.CustomClassMapper
import com.google.firebase.ktx.Firebase
import java.util.concurrent.TimeUnit
import kotlin.reflect.full.declaredMemberFunctions
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.isAccessible

/**
 * Whether the integration tests should run against a local Firestore emulator or the Production
 * environment
 */
private const val CONNECT_TO_EMULATOR = BuildConfig.USE_EMULATOR_FOR_TESTS

/** Default amount of time to wait for a given operation to complete, used by waitFor() helper. */
private const val OPERATION_WAIT_TIMEOUT_MS = 30000L

private const val AUTO_ID_LENGTH = 20

private val AUTO_ID_ALPHABET = ('A'..'Z') + ('a'..'z') + ('0'..'9')

private val settingsWithEmulator = firestoreSettings {}

private val settingsWithoutEmulator = firestoreSettings { setHost("firestore.googleapis.com") }

/**
 * Initializes a [FirebaseFirestore] instance that used for integration test.
 *
 * <p>Note: the Android Emulator treats "10.0.2.2" as its host machine.
 */
val testFirestore: FirebaseFirestore by lazy {
    Firebase.firestore.apply {
        firestoreSettings =
            if (CONNECT_TO_EMULATOR) {
                useEmulator("10.0.2.2", 8080)
                settingsWithEmulator
            } else {
                settingsWithoutEmulator
            }
    }
}

fun <T> waitFor(task: Task<T>, timeoutMS: Long): T {
    return Tasks.await(task, timeoutMS, TimeUnit.MILLISECONDS)
}

fun <T> waitFor(task: Task<T>): T {
    return waitFor(task, OPERATION_WAIT_TIMEOUT_MS)
}

/** Returns a unique id for each Firestore collection used for integration test. */
private fun autoId(): String {
    return (1..AUTO_ID_LENGTH).map { AUTO_ID_ALPHABET.random() }.joinToString("")
}

/**
 * Returns a unique [CollectionReference] identified by the combination of collection name and
 * [autoId] for integration test.
 */
fun testCollection(name: String): CollectionReference {
    return testFirestore.collection("${name}_${autoId()}")
}

/** Returns a [DocumentReference] identified by document name for integration test. */
fun testDocument(name: String): DocumentReference {
    return testCollection("test-collection").document(name)
}

/** Returns a [DocumentReference] for integration test. */
fun testDocument(): DocumentReference {
    return testCollection("test-collection").document()
}

/**
 * Overwrites the document referred to by this [DocumentReference]. If the document does not yet
 * exist, it will be created. If a document already exists, it will be overwritten.
 *
 * @param data The data to write to the document (the data must be a @Serializable Kotlin object).
 * @return A Task that will be resolved when the write finishes.
 */
inline fun <reified T> DocumentReference.setData(data: T): Task<Void> {
    val encodedMap = encodeToMap<T>(data)
    return this.set(encodedMap)
}

/**
 * Returns the contents of the document converted to a Serializable Custom Kotlin Object or null if
 * the document doesn't exist. This method always use the Kotlin Serialization plugin method,and
 * this is a helper function for integration test purpose (to compare Kotlin decode method is the
 * same as Java POJO method).
 *
 * @return The contents of the document in an object of type T or null if the document doesn't
 * exist.
 */
inline fun <reified T> DocumentSnapshot.getData(
    serverTimestampBehavior: DocumentSnapshot.ServerTimestampBehavior =
        DocumentSnapshot.ServerTimestampBehavior.NONE
): T? {
    val mapToBeDecoded = this.getData(serverTimestampBehavior)
    return mapToBeDecoded?.let { decodeFromMap<T>(it, reference) }
}

/**
 * Overwrites the document referred to by this [DocumentReference]. If the document does not yet
 * exist, it will be created. If a document already exists, it will be overwritten. If you pass
 * [SetOptions], the provided data can be merged into an existing document.
 *
 * <p> This method always use the reflection based Java pojo method, and this is a helper function
 * for integration test purpose (to compare Kotlin encode method is the same as Java POJO method).
 *
 * @param data The data to write to the document (e.g. a Map or a POJO containing the desired
 * document contents).
 * @param options An object to configure the set behavior, default is [SetOptions.OVERWRITE]
 * @return A Task that will be resolved when the write finishes.
 */
inline fun <reified T> DocumentReference.setPojoData(
    data: T,
    options: SetOptions? = OVERWRITE_SET_OPTIONS
): Task<Void> {
    return this::class
        .declaredMemberFunctions
        ?.firstOrNull { it.name == "setParsedData" }
        ?.apply { isAccessible = true }
        ?.call(this, data, options) as Task<Void>
}

val OVERWRITE_SET_OPTIONS: SetOptions?
    get() = testSetOptions(false, null)

/**
 * Returns a [SetOptions] from its constructor, this is a helper function for integration test purpose.
 */
private fun testSetOptions(merge: Boolean, field: FieldMask?): SetOptions? =
    SetOptions::class
        .constructors
        .firstOrNull { it.name == "<init>" }
        ?.apply { isAccessible = true }
        ?.call(merge, field)

/**
 * Returns the contents of the document converted to a Custom POJO Object or null if the document
 * doesn't exist. This method always use the reflection based Java pojo method, and this is a helper
 * function for integration test purpose (to compare Kotlin decode method is the same as Java POJO
 * method).
 *
 * @return The contents of the document in an object of type T or null if the document doesn't
 * exist.
 */
inline fun <reified T> DocumentSnapshot.getPojoData(
    serverTimestampBehavior: DocumentSnapshot.ServerTimestampBehavior =
        DocumentSnapshot.ServerTimestampBehavior.NONE
): T? {
    val data = getData(serverTimestampBehavior)
    return CustomClassMapper.convertToCustomClass(data, T::class.java, reference)
}

/**
 * Writes to the document referred to by the provided DocumentReference. If the document does not
 * yet exist, it will be created. The provided data will always override the existing document
 *
 * <p> This method always use the reflection based Java pojo method, and this is a helper function
 * for integration test purpose (to compare Kotlin transaction.set method is the same as Java POJO method)
 *
 * @param documentRef The [DocumentReference] to overwrite.
 * @param data The data to write to the document (e.g. a Map or a POJO containing the desired
 *     document contents).
 * @return This [Transaction] instance. Used for chaining method calls.
 */
inline fun <reified T> Transaction.set(
    documentRef: DocumentReference,
    data: T
): Transaction {
    val firestore =
        Transaction::class
            .memberProperties
            .firstOrNull { it.name == "firestore" }
            ?.apply { isAccessible = true }
            ?.get(this) as FirebaseFirestore
    val transaction =
        Transaction::class
            .memberProperties
            .firstOrNull { it.name == "transaction" }
            ?.apply { isAccessible = true }
            ?.get(this) as com.google.firebase.firestore.core.Transaction
    val firestoreUserDataReader =
        FirebaseFirestore::class
            .memberProperties
            .firstOrNull { it.name == "userDataReader" }
            ?.apply { isAccessible = true }
            ?.get(firestore) as UserDataReader
    val documentRefKey =
        DocumentReference::class
            .memberProperties
            .firstOrNull { it.name == "key" }
            ?.apply { isAccessible = true }
            ?.get(documentRef) as DocumentKey

    val parsed = firestoreUserDataReader.parseSetData(data)
    transaction[documentRefKey] = parsed
    return this
}
